### ISSUE TITLE
[proposal] DistSQL Cache support on TiKV

### ISSUE THREAD

**Issue by blacktear23** [tikv/tikv#4230]
## Feature Request

Add cache on coprocessor level. This can reduce CPU comsumption on same coprocessor request with same data. For support that TiKV need add some feature to let `kv-client` can determine which request can hit the cache.

## Design

First we should update kvproto to support new Request and Response
* Request add field `uint64 match_raft_log_index` it should be optional if absent TiKV should not check raft log index
* Response add field `bool match_raft_log_index` if true means cache can be hited.
* Response add field `bool cacheable` if true means this query can be cache. Because some query function may change each time such as get current time or get random number.
* Response add field `uint64 raft_log_index` store raft log index when the query execute

When TiKV execute a query:
1. TiDB send coprocessor request with match_raft_log_index
2. coprocessor check match_raft_log_index with Snapshot
3. if raft_log_index matched and coprocessor query is cacheable, no need to calculate result just return an empty response and set match_raft_log_index to true
4. if not matched, do calculate and fill response with match_raft_log_index = false and raft_log_index = Snapshot.raft_log_index and set cacheable to true if coprocessor query is cacheable.

pesudo code is like below:

```
fn handle_request() {
    cacheable = this.request.cacheable()
    if (this.request.has_match_raft_log_index() &&
        this.request.get_match_raft_log_index() == this.snapshot.raft_log_index() &&
        cacheable) {
        response = Response::new()
        response.set_match_raft_log_index(true)
        response.set_raft_log_index(this.snapshot.raft_log_index())
        response.set_cacheable(cacheable)
        return response
    } else {
        result = this.executor.execute()
        response = Response::new()
        response.set_data(result.data)
        response.set_match_raft_log_index(false)
        response.set_raft_log_index(this.snapshot.raft_log_index())
        response.set_cacheable(cacheable)
        return response
    }
}
```

## TODO:

1. Update kvproto
2. Update coprocessor let it can get snapshot’s raft log index

**Comment by Connor1996** [comment]
PTAL @breeswish

**Comment by breeswish** [comment]
Could you explain why raft commit index is needed for this feature? We have a plan to extract Coprocessor DAG as a standalone query engine component, which will not have any knowledge about "raft".

**Comment by blacktear23** [comment]
@breeswish yes

**Comment by breeswish** [comment]
Looks like this solution is approachable and can be a cache layer wrapping the existing coprocessor part. How will you handle undetermined expressions like current time and random?

**Comment by blacktear23** [comment]
@breeswish at TiDB part there should be a function to determine whether this query can be cached. And in that function seems we should add some logic to prevent query with those SQL function to use cache. And we can also implements this function on TiKV part instead.

**Comment by breeswish** [comment]
Okay.

**Comment by breeswish** [comment]
Implemented by #6076
